napari/utils/context/_expressions.py

```diff
@@ -0,0 +1,597 @@
+        raise SyntaxError(
+            f"{expr!r} is not a valid expression: ({e})."
+        ) from None
+
+
+def safe_eval(expr: str, context: Context = {}) -> Any:
+    """Safely evaluate `expr` string given `context` dict.
+
+    This lets you evaluate a string expression with broader expression
+    support than `ast.literal_eval`, but much less support than `eval()`.
+    """
+    return parse_expression(expr).eval(context)
+
+
+class Expr(ast.AST, Generic[T]):
+    """Base Expression class providing dunder and convenience methods.
+
+    This class is not meant to be instantiated.
+    Instead, use :func:`parse_expression`, or the `Expr.parse` classmethod
+    to create an expression instance.
+
+    Once created, an expression can be joined with other expressions, or
+    constants.
+
+    Methods
+    -------
+    Outside of the dunder methods providing operator support, key methods are:
+
+    eval(context: dict) -> Any
+        Evaluate this expression in the `context` namespace.
+    parse(expr: str) -> Expr
+        (classmethod) - Parse a string to create a new expression
+    _serialize() -> str
+        serialize the expression to a string
+
+    Examples
+    --------
+    >>> expr = parse_expression('myvar > 5')
+
+    # combine expressions with operators
+    >>> new_expr = expr & parse_expression('v2')
+
+    # nice repr
+    >>> new_expr
+    BoolOp(
+        op=And(),
+        values=[
+            Compare(
+            left=Name(id='myvar', ctx=Load()),
+            ops=[
+                Gt()],
+            comparators=[
+                Constant(value=5)]),
+            Name(id='v2', ctx=Load())])
+
+    # evaluate in some context
+    >>> new_expr.eval(dict(v2='hello!', myvar=8))
+    'hello!'
+
+    # serialize
+    >>> print(new_expr)
+    myvar > 5 and v2
+
+    One reason you might want to use this is to capture named expressions
+    that can be evaluated repeatedly as some underlying context changes.
+
+    .. code-block:: python
+
+        light_is_green = Name[bool]('light_is_green')
+        count = Name[int]('count')
+        is_ready = light_is_green & count > 5
+
+        assert is_ready.eval({'count': 4, 'light_is_green': True}) == False
+        assert is_ready.eval({'count': 7, 'light_is_green': False}) == False
+        assert is_ready.eval({'count': 7, 'light_is_green': True}) == True
+
+    this will also preserve type information:
+    `reveal_type(is_ready())`  # revealed type is `bool`
+    """
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        if type(self).__name__ == 'Expr':
+            raise RuntimeError("Don't instantiate Expr.  Use `Expr.parse`")
+        super().__init__(*args, **kwargs)
+        ast.fix_missing_locations(self)
+
+    def eval(self, context: Context = {}) -> T:
+        """Evaluate this expression with names in `context`"""
+        code = compile(ast.Expression(body=self), '<Expr>', 'eval')
+        try:
+            return eval(code, context)
```

Comment:
@Carreau. curious to get your thoughts on this part.
I'm still using `eval` here, but with the sanitized `Expr` object.  The alternative would be to give each subclass an `eval` method that uses `operator` methods and stuff, rather than compiling this AST to a code object and evaluating it.  I actually had that in a previous version, and could go back to it, but this is way less code, and seems relatively safe.  thoughts?

Reply:
I think that's fine until we get more context on how far we need to go. I'm all for simplicity.